test/ctmrg/partition_function.jl

```diff
@@ -0,0 +1,140 @@
+    NORTH,
+    SOUTH
+
+## Setup
+
+"""
+    ising_exact(beta, J)
+
+[Exact Onsager solution](https://en.wikipedia.org/wiki/Square_lattice_Ising_model#Exact_solution)
+for the 2D classical Ising Model with partition function
+
+```math
+\\mathcal{Z}(\\beta) = \\sum_{\\{s\\}} \\exp(-\\beta H(s)) \\text{ with } H(s) = -J \\sum_{\\langle i, j \\rangle} s_i s_j
+```
+"""
+function classical_ising_exact(; beta=log(1 + sqrt(2)) / 2, J=1.0)
+    K = beta * J
+
+    k = 1 / sinh(2 * K)^2
+    F = quadgk(
+        theta -> log(cosh(2 * K)^2 + 1 / k * sqrt(1 + k^2 - 2 * k * cos(2 * theta))), 0, pi
+    )[1]
+    f = -1 / beta * (log(2) / 2 + 1 / (2 * pi) * F)
+
+    m = 1 - (sinh(2 * K))^(-4) > 0 ? (1 - (sinh(2 * K))^(-4))^(1 / 8) : 0
+
+    E = quadgk(theta -> 1 / sqrt(1 - (4 * k) * (1 + k)^(-2) * sin(theta)^2), 0, pi / 2)[1]
+    e = -J * cosh(2 * K) / sinh(2 * K) * (1 + 2 / pi * (2 * tanh(2 * K)^2 - 1) * E)
+
+    return f, m, e
+end
+
+"""
+    classical_ising(; beta=log(1 + sqrt(2)) / 2)
+
+Implements the 2D classical Ising model with partition function
+
+```math
+\\mathcal{Z}(\\beta) = \\sum_{\\{s\\}} \\exp(-\\beta H(s)) \\text{ with } H(s) = -J \\sum_{\\langle i, j \\rangle} s_i s_j
+```
+"""
+function classical_ising(; beta=log(1 + sqrt(2)) / 2, J=1.0)
+    K = beta * J
+    # Boltzmann weights
+    t = ComplexF64[exp(K) exp(-K); exp(-K) exp(K)]
+    r = eigen(t)
+    nt = r.vectors * sqrt(LinearAlgebra.Diagonal(r.values)) * r.vectors
+
+    # local partition function tensor
+    O = zeros(2, 2, 2, 2)
+    O[1, 1, 1, 1] = 1
+    O[2, 2, 2, 2] = 1
+    @tensor o[-1 -2; -3 -4] := O[3 4; 2 1] * nt[-3; 3] * nt[-4; 4] * nt[-2; 2] * nt[-1; 1]
+
+    # magnetization tensor
+    M = copy(O)
+    M[2, 2, 2, 2] *= -1
+    @tensor m[-1 -2; -3 -4] := M[1 2; 3 4] * nt[-1; 1] * nt[-2; 2] * nt[-3; 3] * nt[-4; 4]
+
+    # bond interaction tensor and energy-per-site tensor
+    e = ComplexF64[-J J; J -J] .* nt
+    @tensor e_hor[-1 -2; -3 -4] :=
+        O[1 2; 3 4] * nt[-1; 1] * nt[-2; 2] * nt[-3; 3] * e[-4; 4]
+    @tensor e_vert[-1 -2; -3 -4] :=
+        O[1 2; 3 4] * nt[-1; 1] * nt[-2; 2] * e[-3; 3] * nt[-4; 4]
+    e = e_hor + e_vert
+
+    # fixed tensor map space for all three
+    TMS = ℂ^2 ⊗ ℂ^2 ← ℂ^2 ⊗ ℂ^2
+
+    return TensorMap(o, TMS), TensorMap(m, TMS), TensorMap(e, TMS)
+end
+
+"""
+Contract a local rank-4 tensor with a given partition function environment.
+"""
+function local_contraction(
+    O::AbstractTensorMap{S,2,2}, env::CTMRGEnv{C,<:CTMRGEdgeTensor{S,2}}
+) where {S,C}
+    return @autoopt @tensor env.corners[NORTHWEST, 1, 1][C_WNW; C_NNW] *
+        env.edges[NORTH, 1, 1][C_NNW D_N; C_NNE] *
+        env.corners[NORTHEAST, 1, 1][C_NNE; C_ENE] *
+        env.edges[EAST, 1, 1][C_ENE D_E; C_ESE] *
+        env.corners[SOUTHEAST, 1, 1][C_ESE; C_SSE] *
+        env.edges[SOUTH, 1, 1][C_SSE D_S; C_SSW] *
+        env.corners[SOUTHWEST, 1, 1][C_SSW; C_WSW] *
+        env.edges[WEST, 1, 1][C_WSW D_W; C_WNW] *
+        O[D_W D_S; D_N D_E]
+end
```

Comment:
This could probably be added to PEPSKit itself, maybe something like:
`expectation_value((i, j) => O, partitionfunction, env)`?

Reply:
I was a bit apprehensive to do this because, while it is an expectation value in the stat mech sense, it is not really the same as the `InfinitePEPS` case where you sandwich an operator between states via a physical level.

But I guess this doesn't matter that much, I'll add it and try to be more generic about the contraction generation so it can handle different kinds of expectation values.

Reply:
I added an `expectation_value` with this signature, only for local (rank-4) `O`s for now though.